src/nonLinPMMH.cpp

```diff
@@ -0,0 +1,146 @@
+	smc::moveset<double> Moveset(fInitialise, fMove, NULL);
+    Sampler.SetResampleParams(ResampleType::MULTINOMIAL, 0.5);
+    Sampler.SetMoveSet(Moveset);
+    Sampler.Initialise();
+	Sampler.IterateUntil(lIterates-1);
+	loglike(0) = Sampler.GetLogNCPath();
+	
+	// Inverse gamma prior
+	logprior(0) = logPrior(theta_prop);
+    
+	double MH_ratio;
+	for (unsigned int i = 1; i<lMCMCits; i++){
+		// RW proposal for parameters
+		theta_prop.sigv = R::rnorm(sigv(i-1),0.15);
+		theta_prop.sigw = R::rnorm(sigw(i-1),0.08);		
+		
+		// Getting a particle filtering estimate of the log likelihood.
+		Sampler.Initialise();
+		Sampler.IterateUntil(lIterates-1);
+		loglike_prop = Sampler.GetLogNCPath();
+		
+		// Inverse gamma prior
+		logprior_prop = logPrior(theta_prop);
+		
+		MH_ratio = exp(loglike_prop - loglike(i-1) + logprior_prop - logprior(i-1));
+		dRand = R::runif(0,1);
+      
+		if (MH_ratio>dRand){
+			sigv(i) = theta_prop.sigv;
+			sigw(i) = theta_prop.sigw;
+			loglike(i) = loglike_prop;
+			logprior(i) = logprior_prop;
+			Rcpp::Rcout  << "ACCEPTED proposal (" << setw(6) << sigv(i) << ", " << setw(6) <<  sigw(i) << ") at iteration " << setw(4) << i+1 << " with loglike " << setw(8) << loglike(i) << std::endl;
+		} else {
+			sigv(i) = sigv(i-1);
+			sigw(i) = sigw(i-1);
+			loglike(i) = loglike(i-1);
+			logprior(i) = logprior(i-1);
+		}
+	}
+		   
+    return Rcpp::DataFrame::create(Rcpp::Named("samples_sigv") = sigv,
+								   Rcpp::Named("samples_sigw") = sigw,
+                                   Rcpp::Named("loglike") = loglike,
+                                   Rcpp::Named("logprior") = logprior);
+  }
+  catch(smc::exception  e) {
+    Rcpp::Rcout << e;
+  }
+  return R_NilValue;          	// to provide a return 
+}
+
+namespace nonLinPMMH {
+	
+/// A function to calculate the log prior for a proposal. The prior for this example is IG(0.01,0.01).
+
+/// \param proposal		The proposed values of the parameters
+double logPrior(const parameters & proposal)
+{
+	return 2*a_prior*log(b_prior)-2*lgamma(a_prior)-(a_prior+1)*log(proposal.sigv)-b_prior/proposal.sigv-(a_prior+1)*log(proposal.sigw)-b_prior/proposal.sigw;
+}
+
+
+/// A function to initialise a particle.
+
+/// \param X			A reference to the empty particle value
+/// \param logweight	A reference to the empty particle log weight
+void fInitialise(double & X, double & logweight)
+{
+	X = R::rnorm(0.0,sqrt(5.0));
+	double mean = pow(X,2)/20.0;
+	logweight = R::dnorm(y(0),mean,theta_prop.sigw,TRUE);
+}
+
+/// The proposal function.
+
+/// \param lTim		The sampler iteration.
+/// \param X			A reference to the current particle value
+/// \param logweight	A reference to the current particle log weight
+void fMove(long lTime, double & X, double & logweight)
+{
+	X = X/2.0 + 25.0*X/(1+pow(X,2)) + 8*cos(1.2*(lTime+1)) + R::rnorm(0.0,theta_prop.sigv);
+	double mean = pow(X,2)/20.0;
+	logweight += R::dnorm(y(lTime),mean,theta_prop.sigw,TRUE);
+}
+
+}
```

Comment:
More trivia: It's conventional (and good practice) to end source files with a new line (some compilers expect the final source line to be terminated with one).